proposals/route-blinding.md

```diff
@@ -0,0 +1,504 @@
+  routes lead to the same recipient (unless of course this information is leaked by higher layers
+  of the protocol, such as using the same `payment_hash` or being generated for the same offer).
+
+### Encrypted data
+
+Route blinding introduces a new TLV field to the onion `tlv_payload`: the `encrypted_data`.
+
+This field is used to carry data coming from the builder of the route that cannot be modified by
+the sender. It needs to contain enough data to let intermediate nodes locate the next node in the
+route (usually a `node_id` or `scid`), and may be extended with additional data in the future. It
+uses ChaCha20-Poly1305 as AEAD scheme.
+
+1. type: 10 (`encrypted_data`)
+2. data:
+    * [`...*byte`:`encrypted_data`]
+
+Once decrypted, the content of this encrypted payload is a TLV stream.
+
+### Creating a blinded route
+
+`N(r)` performs the following steps to create a blinded route:
+
+```text
+Initialization:
+
+  e(0) <- {0;1}^256
+  E(0) = e(0) * G
+
+Blinding:
+
+  For i = 0 to r-1:
+    ss(i) = H(e(i) * N(i)) = H(k(i) * E(i))         // shared secret known only by N(r) and N(i)
+    B(i) = HMAC256("blinded_node_id", ss(i)) * N(i) // Blinded node_id for N(i), private key known only by N(i)
+    rho(i) = HMAC256("rho", ss(i))                  // Key used to encrypt payload for N(i) by N(r)
+    e(i+1) = H(E(i) || ss(i)) * e(i)                // Ephemeral private key, only known by N(r)
+    E(i+1) = H(E(i) || ss(i)) * E(i)                // NB: N(i) must not learn e(i)
+```
+
+Note that this is exactly the same construction as Sphinx, but at each hop we use the shared secret
+to derive a blinded `node_id` for `N(i)` for which the private key will only be known by `N(i)`.
+
+The recipient then creates `encrypted_data(i)` by encrypting application-specific data with
+ChaCha20-Poly1305 using the `rho(i)` key.
+
+To use the blinded route, senders need the following data:
+
+* The real `node_id` of the introduction point `N(0)` (to locate the beginning of the route)
+* The list of blinded `node_id`s: `[B(1),...,B(r)]`
+* The encrypted data for each node: `[encrypted_data(0),...,encrypted_data(r)]`
+* The first blinding ephemeral key: `E(0)`
+
+### Sending to a blinded route
+
+The sender finds a route to the introduction point `N(0)`, and extends it with the blinded route.
+It then creates an onion for that whole route, and includes `E(0)` and `encrypted_data(0)` in the
+onion payload for `N(0)`. It includes `encrypted_data(i)` in the onion payload for `B(i)`.
+
+When `N(0)` receives the onion and decrypts it, it finds `E(0)` in the payload and is able to
+compute the following:
+
+```text
+  ss(0) = H(k(0) * E(0))
+  rho(0) = HMAC256("rho", ss(0))
+  E(1) = H(E(0) || ss(0)) * E(0)
+```
+
+It uses `rho(0)` to decrypt the `encrypted_data(0)` and discovers that `B(1)` is actually `N(1)`.
+It forwards the onion to `N(1)` and includes `E(1)` in a TLV field in the lightning message
+(e.g. in the extension field of an `update_add_htlc` message).
```

Comment:
You need to specify that in https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#adding-an-htlc-update_add_htlc

Reply:
Done in https://github.com/lightning/bolts/pull/765/commits/a78b9740a53ab2197e23f9157e945fe60f1bba27
Once there is rough consensus with a few implementations, we really should update the payment test vectors to cover everything that is specified.